Assets #868 (#874)
plotly/dash: CONTRIBUTING.md

# in your working directory
$ git clone https://github.com/plotly/dash
$ cd dash
$ python3 -m venv .venv/dev
# activate the virtualenv (on windows .venv\dev\scripts\activate)
$ . .venv/dev/bin/activate
# install dash and dependencies
$ pip install -e .[testing,ci]
$ cd dash-renderer
# build renderer bundles, this will build all bundles from source code
# the only true source of npm version is defined in package.json
$ npm run build
# install dash-renderer for development
$ pip install -e .
# you should see both dash and dash-renderer are pointed to local source repos
$ pip list | grep dash

> ```

Author: do we want the gif in the guide too?

Reviewer: I think it's best with just text, no gif.

# in your working directory
$ git clone https://github.com/plotly/dash
$ cd dash
$ python3 -m venv .venv/dev
# activate the virtualenv (on windows .venv\dev\scripts\activate)
$ . .venv/dev/bin/activate
# install dash and dependencies

> $ pip install -e .[testing,ci]  # might need \ to espcae []

Reviewer: ~escpae~ escape

But, is that true? Is this a concern on a particular OS or something?

Author: on zsh, you need \ to escape

Reviewer: OK, so let's change to 
```suggestion
$ pip install -e .[testing,ci]  # in some shells you need \ to escape []
```

$ cd dash-renderer
# build renderer bundles, this will build all bundles from source code
# the only true source of npm version is defined in package.json
$ npm run build
# install dash-renderer for development
$ pip install -e .
# you should see both dash and dash-renderer are pointed to local source repos
$ pip list | grep dash
```

### Dash-Renderer Beginner Guide

> `Dash Renderer`  was a separate Dash project. It was merged into main  `Dash`  repository as part of 1.0 release. The frontend Dash, at its core, is driven by the renderer to handle React Components layout and backend event handling.

Reviewer: ```suggestion
`Dash Renderer` began as a separate repository. It was merged into the main  `Dash` repository as part of the 1.0 release. It is the common frontend for all Dash backends (**R** and **Python**), and manages React Component layout and backend event handling.
```

> If you want to contribute or simply dig deeper into Dash. We encourage you to play and taste it. This is the most efficient way to learn and understand everything under the hood.

Reviewer: ```suggestion
If you want to contribute or simply dig deeper into Dash, we encourage you to play and taste it. This is the most efficient way to learn and understand everything under the hood.
```

> For contributors who have purely  **Python**  or  **R**  background. This section might help you understand more details about developing and debugging in Javascript world.

Reviewer: ```suggestion
For contributors with a primarily  **Python**  or  **R**  background, this section might help you understand more details about developing and debugging in Javascript world.
```

> After Dash 1.2, The renderer bundle and its peer dependencies can be packed and generated from the source code. The only version of the truth is defined in  `dash-renderer\package.json`  file. A build tool  `renderer`, which is a tiny Python script defined as a Dash entry point, has few  commands like:

Reviewer: ```suggestion
As of Dash 1.2, The renderer bundle and its peer dependencies can be packed and generated from the source code. Dependency files are no longer included in this repository, they are downloaded when needed based on the versions listed in `dash-renderer/package.json`. A build tool `renderer`, which is a tiny Python script installed by Dash as a command-line tool, has a few commands:
```

> 1.  `renderer npm`  installs all the npm modules using this  `package.json`  files. Note that the  `package-lock.json`  file is the computed reference product for the versions defined with tilde(~) or caret(^) syntax in  **npm**

Reviewer: ```suggestion
1.  `renderer npm`  installs all the npm modules using this  `package.json`  files. Note that the  `package-lock.json`  file is the computed reference product for the versions defined with tilde(~) or caret(^) syntax in  **npm**. This is called by `npm run build`, or you can also call it directly if you add or update a dependency module.
```

Author: it's slightly safer for people not family with js to call with `renderer` as it will handle the location with `os.chdir()`.

Reviewer: Good point - I just wanted to give people some idea when to use each of these commands. So perhaps:

> Call this if you add or update a dependency module. `npm run build` calls this and `renderer bundle`, but you can only call it from the `dash-renderer` directory.

?

> 2.  `renderer bundles`  parses the locked version JSON, copies all the peer dependencies into dash_renderer folder, bundles the renderer assets, and generates an `__init__.py`  to map all the resources

Reviewer: ```suggestion
2.  `renderer bundles`  parses the locked version JSON, copies all the peer dependencies into dash_renderer folder, bundles the renderer assets, and generates an `__init__.py`  to map all the resources. This is called by `npm run build`, or you can call it directly if the renderer code has changed.
```

> 3.  `renderer digest {renderer version}`  computes the content hash of each asset in  `dash_renderer`  folder, prints out the result in logs, and dumps into a JSON file  `digest.json`

Reviewer: What should we say about when to use this? Just that we use it internally to test that the build has not changed?

Author: yes, something the assets bundle check
I extracted this out as a single job, so we can reuse it when we want to do some simple compare from the hash

> 4.  `renderer watch` runs the webpack in watch mode, so any source code change triggers a rebuild

Reviewer: ```suggestion
4.  `renderer watch` runs webpack in watch mode, so any source code change triggers a rebuild. Use this if you are actively updating renderer code and you want to test your changes immediately.
```

2.  `renderer bundles` parses the locked version JSON, copies all the peer dependencies into dash_renderer folder, bundles the renderer assets, and generates an `__init__.py` to map all the resources. There are also a list of helpful `scripts` defined in `package.json` you might need to do some handy tasks like linting, syntax format with prettier, etc.
3.  `renderer digest` computes the content hash of each asset in `dash_renderer` folder, prints out the result in logs, and dumps into a JSON file `digest.json`. Use this when you have a doubt about the current assets in `dash_renderer`, and compare it with previous result in one shot by this command.
4.  `renderer watch` runs the webpack in watch mode, so any source code change triggers a rebuild. Use this if you are actively updating renderer code and you want to test your changes immediately.

When a change in renderer code doesn't reflect in your browser as expected, this could be: confused bundle generation, caching issue in a browser, python package not in `editable` mode, etc. The new tool reduces the risk of bundle assets by adding the digest to help compare asset changes.

## Git

> Use the [GitHub flow](https://guides.github.com/introduction/flow/) when proposing contributions to this repository (i.e. create a feature branch and submit a PR against the default branch).

Reviewer: FWIW I think I generally prefer links inline like this, but it did work the way it was, with the link defined at the end of the doc.
